Deleting the cluster results in panic

```E0502 12:21:39.508378   21941 runtime.go:64] Observed a panic: &runtime.TypeAssertionError{interfaceString:"interface {}", concreteString:"cache.DeletedFinalStateUnknown", assertedString:"*awstpr.CustomObject", missingMethod:""} (interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *awstpr.CustomObject)
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:70
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:63
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:49
/home/vador/.gimme/versions/go1.8.linux.amd64/src/runtime/asm_amd64.s:514
/home/vador/.gimme/versions/go1.8.linux.amd64/src/runtime/panic.go:489
/home/vador/.gimme/versions/go1.8.linux.amd64/src/runtime/iface.go:172
/home/vador/go/src/github.com/giantswarm/aws-operator/service/create/service.go:570
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/controller.go:187
<autogenerated>:52
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/controller.go:257
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:420
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/controller.go:131
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/controller.go:102
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/wait/wait.go:96
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/wait/wait.go:97
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/wait/wait.go:52
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/controller.go:102
/home/vador/go/src/github.com/giantswarm/aws-operator/service/create/service.go:846
/home/vador/.gimme/versions/go1.8.linux.amd64/src/sync/once.go:44
/home/vador/go/src/github.com/giantswarm/aws-operator/service/create/service.go:847
/home/vador/go/src/github.com/giantswarm/aws-operator/service/service.go:149
/home/vador/.gimme/versions/go1.8.linux.amd64/src/sync/once.go:44
/home/vador/go/src/github.com/giantswarm/aws-operator/service/service.go:150
/home/vador/.gimme/versions/go1.8.linux.amd64/src/runtime/asm_amd64.s:2197
E0502 12:21:39.509299   21941 runtime.go:64] Observed a panic: &runtime.TypeAssertionError{interfaceString:"interface {}", concreteString:"cache.DeletedFinalStateUnknown", assertedString:"*awstpr.CustomObject", missingMethod:""} (interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *awstpr.CustomObject)
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:70
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:63
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:49
/home/vador/.gimme/versions/go1.8.linux.amd64/src/runtime/asm_amd64.s:514
/home/vador/.gimme/versions/go1.8.linux.amd64/src/runtime/panic.go:489
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:56
/home/vador/.gimme/versions/go1.8.linux.amd64/src/runtime/asm_amd64.s:514
/home/vador/.gimme/versions/go1.8.linux.amd64/src/runtime/panic.go:489
/home/vador/.gimme/versions/go1.8.linux.amd64/src/runtime/iface.go:172
/home/vador/go/src/github.com/giantswarm/aws-operator/service/create/service.go:570
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/controller.go:187
<autogenerated>:52
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/controller.go:257
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:420
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/controller.go:131
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/controller.go:102
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/wait/wait.go:96
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/wait/wait.go:97
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/wait/wait.go:52
/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/controller.go:102
/home/vador/go/src/github.com/giantswarm/aws-operator/service/create/service.go:846
/home/vador/.gimme/versions/go1.8.linux.amd64/src/sync/once.go:44
/home/vador/go/src/github.com/giantswarm/aws-operator/service/create/service.go:847
/home/vador/go/src/github.com/giantswarm/aws-operator/service/service.go:149
/home/vador/.gimme/versions/go1.8.linux.amd64/src/sync/once.go:44
/home/vador/go/src/github.com/giantswarm/aws-operator/service/service.go:150
/home/vador/.gimme/versions/go1.8.linux.amd64/src/runtime/asm_amd64.s:2197
panic: interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *awstpr.CustomObject [recovered]
	panic: interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *awstpr.CustomObject [recovered]
	panic: interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *awstpr.CustomObject

goroutine 6 [running]:
github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:56 +0x126
panic(0x1a3ed60, 0xc4208aa940)
	/home/vador/.gimme/versions/go1.8.linux.amd64/src/runtime/panic.go:489 +0x2cf
github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:56 +0x126
panic(0x1a3ed60, 0xc4208aa940)
	/home/vador/.gimme/versions/go1.8.linux.amd64/src/runtime/panic.go:489 +0x2cf
github.com/giantswarm/aws-operator/service/create.(*Service).Boot.func1.2(0x1ac1be0, 0xc420876860)
	/home/vador/go/src/github.com/giantswarm/aws-operator/service/create/service.go:570 +0x514a
github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnDelete(0xc4203ae390, 0x0, 0xc4203ae3a0, 0x1ac1be0, 0xc420876860)
	/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/controller.go:187 +0x49
github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnDelete(0xc4203102a0, 0x1ac1be0, 0xc420876860)
	<autogenerated>:52 +0x73
github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache.NewInformer.func1(0x1a502a0, 0xc4208a1160, 0x1a502a0, 0xc4208a1160)
	/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/controller.go:257 +0x516
github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop(0xc4201ba370, 0xc42032cd20, 0x0, 0x0, 0x0, 0x0)
	/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:420 +0x218
github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache.(*Controller).processLoop(0xc42031c380)
	/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/controller.go:131 +0x3c
github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache.(*Controller).(github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache.processLoop)-fm()
	/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/controller.go:102 +0x2a
github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/wait.JitterUntil.func1(0xc4205f9dd8)
	/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/wait/wait.go:96 +0x5e
github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/wait.JitterUntil(0xc420397dd8, 0x3b9aca00, 0x0, 0x18fd001, 0x0)
	/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/wait/wait.go:97 +0xad
github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/wait.Until(0xc4205f9dd8, 0x3b9aca00, 0x0)
	/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/pkg/util/wait/wait.go:52 +0x4d
github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache.(*Controller).Run(0xc42031c380, 0x0)
	/home/vador/go/src/github.com/giantswarm/aws-operator/vendor/k8s.io/client-go/tools/cache/controller.go:102 +0x1c7
github.com/giantswarm/aws-operator/service/create.(*Service).Boot.func1()
	/home/vador/go/src/github.com/giantswarm/aws-operator/service/create/service.go:846 +0x3e4
sync.(*Once).Do(0xc4202698f0, 0xc4205eb738)
	/home/vador/.gimme/versions/go1.8.linux.amd64/src/sync/once.go:44 +0xbe
github.com/giantswarm/aws-operator/service/create.(*Service).Boot(0xc420269880)
	/home/vador/go/src/github.com/giantswarm/aws-operator/service/create/service.go:847 +0x55
github.com/giantswarm/aws-operator/service.(*Service).Boot.func1()
	/home/vador/go/src/github.com/giantswarm/aws-operator/service/service.go:149 +0x2d
sync.(*Once).Do(0xc420367310, 0xc4205eb7b8)
	/home/vador/.gimme/versions/go1.8.linux.amd64/src/sync/once.go:44 +0xbe
github.com/giantswarm/aws-operator/service.(*Service).Boot(0xc420367300)
	/home/vador/go/src/github.com/giantswarm/aws-operator/service/service.go:150 +0x55
created by main.main.func1
	/home/vador/go/src/github.com/giantswarm/aws-operator/main.go:104 +0x449
```

--- a/service/create/service.go
+++ b/service/create/service.go
@@ -567,7 +567,29 @@ func (s *Service) Boot() {
 				},
 				DeleteFunc: func(obj interface{}) {
 					// TODO(nhlfr): Move this to a separate operator.
-					cluster := *obj.(*awstpr.CustomObject)
+
+					// We can receive an instance of awstpr.CustomObject or cache.DeletedFinalStateUnknown.
+					// We need to assert the type properly and log an error when we cannot do that.
+					// Also, the cache.DeleteFinalStateUnknown object can contain the proper CustomObject,
+					// but doesn't have to.
+					// https://github.com/kubernetes/client-go/blob/7ba6be594966f4bec08a57a60c855d17a9f7000a/tools/cache/delta_fifo.go#L674-L677
+					var cluster awstpr.CustomObject
+					clusterPtr, ok := obj.(*awstpr.CustomObject)
+					if ok {
+						cluster = *clusterPtr
+					} else {
+						deletedObj, ok := obj.(cache.DeletedFinalStateUnknown)
+						if !ok {
+							s.logger.Log("error", "received unknown type of third-party object")
+							return
+						}
+						clusterPtr, ok := deletedObj.Obj.(*awstpr.CustomObject)
+						if !ok {
+							s.logger.Log("error", "received the proper delete request, but the type of third-party object is unknown")
+							return
+						}
+						cluster = *clusterPtr
+					}
 
 					if err := s.deleteClusterNamespace(cluster.Spec.Cluster); err != nil {
 						s.logger.Log("error", "could not delete cluster namespace:", err)

